db/pebbledb/pebledb.go

```diff
@@ -61,39 +62,53 @@ func iterate(reader pebble.Reader, prefix []byte, callback func(k, v []byte) boo
 }
 
 // Get implements the db.WriteTx.Get interface method
-func (tx WriteTx) Get(k []byte) ([]byte, error) {
+func (tx *WriteTx) Get(k []byte) ([]byte, error) {
 	return get(tx.batch, k)
 }
 
-func (tx WriteTx) Iterate(prefix []byte, callback func(k, v []byte) bool) (err error) {
+func (tx *WriteTx) Iterate(prefix []byte, callback func(k, v []byte) bool) (err error) {
 	return iterate(tx.batch, prefix, callback)
 }
 
 // Set implements the db.WriteTx.Set interface method
-func (tx WriteTx) Set(k, v []byte) error {
+func (tx *WriteTx) Set(k, v []byte) error {
 	return tx.batch.Set(k, v, nil)
 }
 
 // Delete implements the db.WriteTx.Delete interface method
-func (tx WriteTx) Delete(k []byte) error {
+func (tx *WriteTx) Delete(k []byte) error {
 	return tx.batch.Delete(k, nil)
 }
 
 // Apply implements the db.WriteTx.Apply interface method
-func (tx WriteTx) Apply(other db.WriteTx) (err error) {
-	otherPebble := db.UnwrapWriteTx(other).(WriteTx)
+func (tx *WriteTx) Apply(other db.WriteTx) (err error) {
+	otherPebble := db.UnwrapWriteTx(other).(*WriteTx)
 	return tx.batch.Apply(otherPebble.batch, nil)
 }
 
 // Commit implements the db.WriteTx.Commit interface method
-func (tx WriteTx) Commit() error {
-	return tx.batch.Commit(nil)
+func (tx *WriteTx) Commit() error {
+	if tx.batch == nil {
+		return fmt.Errorf("cannot commit pebble tx: already committed or discarded")
+	}
+	err := tx.batch.Commit(nil)
+	tx.batch = nil
+	return err
 }
 
 // Discard implements the db.WriteTx.Discard interface method
-func (tx WriteTx) Discard() {
+func (tx *WriteTx) Discard() {
+	if tx.batch == nil {
+		// Silently allow discarding twice or after a commit,
+		// since it can help for the sake of defers.
+		// Note that upstream pebbledb doesn't really allow calling Close twice;
+		// the object gets put in a shared pool, so a later call to Close races
+		// with another goroutine having already reused the same object from the pool.
+		return
+	}
 	// Close returns an error, but here in the Discard context is omitted
 	tx.batch.Close()
+	tx.batch = nil
```

Comment:
Isn't this still racy? Since it is not atomic and there is no lock, two Discard() calls could be done in parallel and the second would call tx.batch.Close() of a nil object.

Reply:
The race in master is not because of two goroutines calling Discard concurrently on the same tx. We have never supported concurrent calls on the same tx like that, and there's no reason why we should - a single tx is meant to be for a sequence of operations done by a single goroutine.

The race happens when one goroutine grabs one tx (T1) and discards it twice, and another goroutine does the same on another tx (T2). Because pebble uses a global pool of transactions, T1 and T2 might actually be shared after the first discard, since pebble's close method puts the batch back in the pool - so then the second discard can be racy.

Not sure if that made any sense - but the kind of race you mean was never a problem and shouldn't be a problem.